Don't wait 8 minutes to suspend background tabs on iOS

bnham opened the pull request:
#### b0259bf3dba0415fb969fba15169c1ec7646917a
<pre>
Don&apos;t wait 8 minutes to suspend background tabs on iOS
<a href="https://bugs.webkit.org/show_bug.cgi?id=265064">https://bugs.webkit.org/show_bug.cgi?id=265064</a>
<a href="https://rdar.apple.com/118578232">rdar://118578232</a>

Reviewed by Chris Dumez.

Disable the NearSuspended state, which allows background WebContent processes to suspend immediately
rather than after 8 minutes.

To do this, we need to fix the heuristic in `prepareToDropLastAssertion`, which fires just before a
process task_suspends. Previously it was generally called after a process ended up in the process
cache. Now that we are calling it earlier, we want it to prevent clearing out the caches from
processes that might end up in the process cache in the future, so we have to consult
`canBeAddedToWebProcessCache` about that.

Another change is that the `SetIsInProcessCache` message might be sent while the process is
task_suspended. To fix this, change that message to one with an async reply handler so that it takes
out a background activity while the IPC is being handled.

Finally, the `WebProcess::releaseMemory` IPC handler was often causing the process to completely exit,
since it invokes the memory pressure handler, and the memory pressure handler exits if there are no
active pages in the process. This seems undesired (if the UIProcess really wanted to terminate the
WebProcess, it could just send the `AuxiliaryProcess::ShutDown` message). So we now suppress the
memory pressure handler called by `WebProcess::releaseMemory` from exiting the process.

* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::setIsInProcessCache):
(WebKit::WebProcessProxy::prepareToDropLastAssertion):
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::initializeWebProcess):
(WebKit::WebProcess::setIsInProcessCache):
(WebKit::WebProcess::releaseMemory):
(WebKit::WebProcess::prepareToSuspend):
* Source/WebKit/WebProcess/WebProcess.h:
* Source/WebKit/WebProcess/WebProcess.messages.in:

Canonical link: <a href="https://commits.webkit.org/272937@main">https://commits.webkit.org/272937@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/12c5bf1fb55c395ed697afac19bfdcc2e01dd5ac

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/33637 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/12411 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/35566 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/36255 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/30514 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/14784 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/9568 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/29607 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/34113 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/10463 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/29992 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/9140 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/9226 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/30000 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/37581 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/28665 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/30531 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/30332 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/35364 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/33593 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/9369 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/7323 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/33249 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/11147 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/40106 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7780 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/9953 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/8398 "Passed tests") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/10141 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/UIProcess/WebProcessProxy.cpp:
```diff
@@ -390,7 +390,7 @@ void WebProcessProxy::setIsInProcessCache(bool value, WillShutDown willShutDown)
         return;
 
-    send(Messages::WebProcess::SetIsInProcessCache(m_isInProcessCache), 0);
+    sendWithAsyncReply(Messages::WebProcess::SetIsInProcessCache(m_isInProcessCache), []() { });
```
Maybe add a comment to explain that we're using sendWithAsyncReply() to make sure the destination process gets woken up long enough to process the IPC.

cdumez commented on Source/WebKit/WebProcess/WebProcess.cpp:
```diff
@@ -687,6 +687,7 @@ void WebProcess::setIsInProcessCache(bool isInProcessCache)
 
     updateProcessName(IsInProcessInitialization::No);
     accessibilityRelayProcessSuspended(isInProcessCache);
+    completionHandler();
```
You're failing to call the completion handler if the `#else` case. We should probably just move this out of the `#if / #else` scope.

cdumez reviewed (approved):
LGTM

bnham committed:
Don't wait 8 minutes to suspend background tabs on iOS
https://bugs.webkit.org/show_bug.cgi?id=265064
rdar://118578232

Reviewed by Chris Dumez.

Disable the NearSuspended state, which allows background WebContent processes to suspend immediately
rather than after 8 minutes.

To do this, we need to fix the heuristic in `prepareToDropLastAssertion`, which fires just before a
process task_suspends. Previously it was generally called after a process ended up in the process
cache. Now that we are calling it earlier, we want it to prevent clearing out the caches from
processes that might end up in the process cache in the future, so we have to consult
`canBeAddedToWebProcessCache` about that.

Another change is that the `SetIsInProcessCache` message might be sent while the process is
task_suspended. To fix this, change that message to one with an async reply handler so that it takes
out a background activity while the IPC is being handled.

Finally, the `WebProcess::releaseMemory` IPC handler was often causing the process to completely exit,
since it invokes the memory pressure handler, and the memory pressure handler exits if there are no
active pages in the process. This seems undesired (if the UIProcess really wanted to terminate the
WebProcess, it could just send the `AuxiliaryProcess::ShutDown` message). So we now suppress the
memory pressure handler called by `WebProcess::releaseMemory` from exiting the process.

* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::setIsInProcessCache):
(WebKit::WebProcessProxy::prepareToDropLastAssertion):
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::initializeWebProcess):
(WebKit::WebProcess::setIsInProcessCache):
(WebKit::WebProcess::releaseMemory):
(WebKit::WebProcess::prepareToSuspend):
* Source/WebKit/WebProcess/WebProcess.h:
* Source/WebKit/WebProcess/WebProcess.messages.in:

Canonical link: https://commits.webkit.org/272937@main
